index.js

```diff
@@ -16,117 +16,123 @@ function helmetPlugin (fastify, options, next) {
+    const { helmet: routeOptions } = request.context.config
+
+    if (typeof routeOptions !== 'undefined') {
+      const { enableCSPNonces: enableRouteCSPNonces, skipRoute, ...helmetRouteConfiguration } = routeOptions
+
+      if (skipRoute === true) {
+        // If helmet route option is set to `false` we skip the route
+      } else {
+        // If route helmet options are set they overwrite the global helmet configuration
+        const mergedHelmetConfiguration = Object.assign(Object.create(null), globalConfiguration, helmetRouteConfiguration)
+
+        buildHelmetOnRoutes(request, reply, mergedHelmetConfiguration, enableRouteCSPNonces)
+      }
 
-function buildCSPNonce (fastify, configuration) {
-  const cspDirectives = configuration.contentSecurityPolicy
-    ? configuration.contentSecurityPolicy.directives
-    : helmet.contentSecurityPolicy.getDefaultDirectives()
-  const cspReportOnly = configuration.contentSecurityPolicy
-    ? configuration.contentSecurityPolicy.reportOnly
-    : undefined
-
-  return function (request, reply, next) {
-    if (!fastify.hasReplyDecorator('cspNonce')) {
-      fastify.decorateReply('cspNonce', null)
+      return next()
+    } else if (isGlobal) {
+      // if the plugin is set globally (meaning that all the routes will be decorated)
+      // As the endpoint, does not have a custom helmet configuration, use the global one.
+      buildHelmetOnRoutes(request, reply, globalConfiguration, enableCSPNonces)
+    } else {
+      // if the plugin is not global we can skip the route
     }
 
-    // prevent object reference: https://github.com/fastify/fastify-helmet/issues/118
-    const directives = { ...cspDirectives }
+    return next()
+  })
+}
 
-    // create csp nonce
+async function replyDecorators (request, reply, configuration, enableCSP) {
+  if (enableCSP) {
     reply.cspNonce = {
       script: randomBytes(16).toString('hex'),
       style: randomBytes(16).toString('hex')
     }
+  }
 
-    // push nonce to csp
-    // allow both script-src or scriptSrc syntax
-    const scriptKey = Array.isArray(directives['script-src']) ? 'script-src' : 'scriptSrc'
-    directives[scriptKey] = Array.isArray(directives[scriptKey]) ? [...directives[scriptKey]] : []
-    directives[scriptKey].push(`'nonce-${reply.cspNonce.script}'`)
-    // allow both style-src or styleSrc syntax
-    const styleKey = Array.isArray(directives['style-src']) ? 'style-src' : 'styleSrc'
-    directives[styleKey] = Array.isArray(directives[styleKey]) ? [...directives[styleKey]] : []
-    directives[styleKey].push(`'nonce-${reply.cspNonce.style}'`)
+  reply.helmet = function (opts) {
+    const helmetConfiguration = opts
+      ? Object.assign(Object.create(null), configuration, opts)
+      : configuration
 
-    const cspMiddleware = helmet.contentSecurityPolicy({ directives, reportOnly: cspReportOnly })
-    cspMiddleware(request.raw, reply.raw, next)
+    return helmet(helmetConfiguration)(request.raw, reply.raw, done)
```

Comment:
I do not think it is the correct approach for the done method. 
The hook may end before throwing error.

The `done` method should inherit from the `onRequest` hook, there are problem in previous version because it do not terminate the the script in the correct position and it run the `done` method twice.

```suggestion
    return helmet(helmetConfiguration)(request.raw, reply.raw, done)
```